tools/wake-format/emitter.cpp

```diff
@@ -524,19 +527,307 @@ wcl::doc Emitter::walk_ascribe(ctx_t ctx, CSTElement node) {
+
+  if (!(op.id() == CST_OP && op.firstChildElement().id() == collect_over.id() &&
+        op.firstChildElement().fragment().segment().str() ==
+            collect_over.fragment().segment().str())) {
+    return {node};
+  }
+
+  std::vector<CSTElement> collect = {left};
+  auto right_collect = collect_right_binary(collect_over, right);
+
+  collect.insert(collect.end(), right_collect.begin(), right_collect.end());
+
+  return collect;
+}
+
+size_t count_allowed_newlines(const token_traits_map_t& traits,
+                              const std::vector<CSTElement>& parts) {
+  assert(parts.size() >= 2);
+
+  CSTElement first_token = parts[0].firstChildElement();
+
+  CSTElement last_token;
+  {
+    IsWSNLCPredicate is_wsnlc;
+    CSTElement curr_rhs = parts.back().firstChildElement();
+    CSTElement next_rhs = curr_rhs;
+    next_rhs.nextSiblingElement();
+
+    while (!next_rhs.empty()) {
+      while (!next_rhs.empty() && is_wsnlc(next_rhs)) {
+        next_rhs.nextSiblingElement();
+      }
+      if (next_rhs.empty()) {
+        last_token = curr_rhs;
+      } else {
+        curr_rhs = next_rhs;
+        next_rhs.nextSiblingElement();
+      }
+    }
+    last_token = curr_rhs;
+  }
+
+  size_t allowed = 0;
+
+  auto first_it = traits.find(first_token);
+  if (first_it != traits.end()) {
+    allowed += first_it->second.before_bound.size();
+  }
+
+  auto last_it = traits.find(last_token);
+  if (last_it != traits.end()) {
+    allowed += last_it->second.after_bound.size();
+  }
+
+  return allowed;
+}
+
+bool min_doc_height(const wcl::doc& lhs, const wcl::doc& rhs) {
+  return lhs->height() < rhs->height();
+}
+
+bool min_doc_max_width(const wcl::doc& lhs, const wcl::doc& rhs) {
+  return lhs->max_width() < rhs->max_width();
+}
+
+// TODO: this is far from fully correct
+bool is_op_left_assoc(const CSTElement& op) { return op.id() == TOKEN_OP_OR; }
+
+wcl::doc select_best_choice(std::vector<wcl::optional<wcl::doc>> choices) {
+  std::vector<wcl::doc> lte_fmt = {};
+  std::vector<wcl::doc> gt_fmt = {};
+  int i = 0;
+
+  for (auto choice_opt : choices) {
+    i++;
+    if (!choice_opt) {
+      continue;
+    }
+    auto choice = *choice_opt;
+
+    if (choice->max_width() > MAX_COLUMN_WIDTH) {
+      gt_fmt.push_back(std::move(choice));
+    } else {
+      lte_fmt.push_back(std::move(choice));
+    }
+  }
+
+  if (lte_fmt.size() > 0) {
+    auto min_height = std::min_element(lte_fmt.begin(), lte_fmt.end(), min_doc_height);
+    return *min_height;
+  }
+
+  auto min_width = std::min_element(gt_fmt.begin(), gt_fmt.end(), min_doc_max_width);
+  return *min_width;
+}
+
+wcl::doc post_binop_spacing(CSTElement over, const token_traits_map_t& traits,
+                            const wcl::doc& fmt_binop, const wcl::doc_builder& builder, ctx_t ctx) {
+  if (fmt_binop->has_newline()) {
+    return fmt().freshline().compose(ctx.sub(builder), over, traits);
+  }
+  return fmt().space().compose(ctx.sub(builder), over, traits);
+}
```

Comment:
Can you document what this is doing?

Also please static all the static thingies